datafusion/core/src/physical_plan/memory.rs

```diff
@@ -223,15 +245,365 @@ impl RecordBatchStream for MemoryStream {
     }
 }
 
+/// Execution plan for writing record batches to an in-memory table.
+pub struct MemoryWriteExec {
```

Comment:
> But I thought the consensus was to have a generic ExecutionPlan for writing data, rather than one that was specific to MemTable

Are you sure about this? My understanding was going through the `ExecutionPlan` route for each writer, not a single plan for everything. There is a separate `Exec` on the read side for every `TableProvider`, why would we not follow the same pattern?

Anyways, I asked @metesynnada to experiment with the single plan idea, but he wasn't able to make progress on it -- he will comment here with his findings.

Reply:
I agree that a single plan is likely not tractable, it will be highly specific to the `TableProvider` in question. What I had meant was simply not returning anything and leaving everything an internal implementation detail of the `TableProvider` from how it executes the provided `ExecutionPlan` through to how that is exposed. Not only would this make this PR significantly simpler, obviating the need for any of the ExecutionPlan or Stream shenanigans, but would provide the most flexibility for table implementations that have more complex scheduling or transaction handling requirements. Exposing `ExecutionPlan` is a pretty strong API commitment, and I _personally_ am not confident this won't be painting ourselves into a corner, but if the consensus is otherwise :shrug:

Reply:
> All in all, having a distinct ExecutionPlan to handle writes for each data source mirrors the behavior on the read side and seems much more natural. What do you think?

I guess I don't fully understand why adding 800 lines of `MemTable` specific ExecutionPlan  that seems to add no new functionality is an improvement

Not only only this this  pattern seem to  overly complicate this PR, but it sets us on a course for every other table source to have to do the same (mostly repeated) things.

Perhaps there is a future use case I don't understand.

If you guys feel strongly that this is the right and maintainable solution going forward I will defer to your judgement, but from where I see it just adds complexity / technical debt for no reason.

Reply:
This is simply a baby step for implementing things like `COPY TO`s, `INSERT INTO`s and functionality like that. In practical cases, these things will write their outputs through a `TableProvider` like a `ListingTable`. How do we test what that will look like and iterate on such functionality? Basically, this is the simplest writer that one can perform end-to-end (read-transform-write) experiments with, its main intention is for anyone to use as a basis/example for implementing their write operators. I hope it makes a little more sense now.

> If you guys feel strongly that this is the right and maintainable solution going forward I will defer to your judgement, but from where I see it just adds complexity / technical debt for no reason.

Thank you -- rest assured we will be maintaining and helping with related work going forward.